imap_processing/ultra/l1c/ultra_l1c_pset_bins.py

```diff
@@ -44,11 +48,221 @@ def build_spatial_bins(spacing: float = 0.5) -> tuple[np.ndarray, np.ndarray]:
     """
     # Azimuth bins from 0 to 360 degrees.
     az_bin_edges = np.arange(0, 360 + spacing, spacing)
+    az_bin_midpoints = az_bin_edges[:-1] + spacing / 2  # Midpoints between edges
 
     # Elevation bins from -90 to 90 degrees.
     el_bin_edges = np.arange(-90, 90 + spacing, spacing)
+    el_bin_midpoints = el_bin_edges[:-1] + spacing / 2  # Midpoints between edges
+
+    return az_bin_edges, el_bin_edges, az_bin_midpoints, el_bin_midpoints
+
+
+def cartesian_to_spherical(
+    v: tuple[np.ndarray, np.ndarray, np.ndarray],
+) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
+    """
+    Convert cartesian coordinates to spherical coordinates.
+
+    Parameters
+    ----------
+    v : tuple[np.ndarray, np.ndarray, np.ndarray]
+        The x,y,z-components of the velocity vector.
+
+    Returns
+    -------
+    az : np.ndarray
+        The azimuth angles in degrees.
+    el : np.ndarray
+        The elevation angles in degrees.
+    r : np.ndarray
+        The radii, or magnitudes, of the vectors.
+    """
+    vx, vy, vz = np.hsplit(v, 3)
+    vx, vy, vz = vx.flatten(), vy.flatten(), vz.flatten()  # Flatten the arrays
+
+    # Magnitude of the velocity vector
+    magnitude_v = np.sqrt(vx**2 + vy**2 + vz**2)
+
+    vhat_x = -vx / magnitude_v
+    vhat_y = -vy / magnitude_v
+    vhat_z = -vz / magnitude_v
+
+    # Convert from cartesian to spherical coordinates (azimuth, elevation)
+    # Radius (magnitude)
+    r = np.sqrt(vhat_x**2 + vhat_y**2 + vhat_z**2)
+
+    # Elevation angle (angle from the z-axis, range: [-pi/2, pi/2])
+    el = np.arcsin(vhat_z / r)
+
+    # Azimuth angle (angle in the xy-plane, range: [0, 2*pi])
+    az = np.arctan2(vhat_y, vhat_x)
+
+    # Ensure azimuth is from 0 to 2PI
+    az = az % (2 * np.pi)
+
+    return np.degrees(az), np.degrees(el), r
+
+
+def get_histogram(
+    v: tuple[np.ndarray, np.ndarray, np.ndarray],
+    energy: np.ndarray,
+    az_bin_edges: np.ndarray,
+    el_bin_edges: np.ndarray,
+    energy_bin_edges: np.ndarray,
+) -> tuple[NDArray, NDArray]:
+    """
+    Compute a 3D histogram of the particle data.
+
+    Parameters
+    ----------
+    v : tuple[np.ndarray, np.ndarray, np.ndarray]
+        The x,y,z-components of the velocity vector.
+    energy : np.ndarray
+        The particle energy.
+    az_bin_edges : np.ndarray
+        Array of azimuth bin boundary values.
+    el_bin_edges : np.ndarray
+        Array of elevation bin boundary values.
+    energy_bin_edges : np.ndarray
+        Array of energy bin edges.
+
+    Returns
+    -------
+    hist : np.ndarray
+        A 3D histogram array.
+    non_zero_counts : np.ndarray
+        An array of non-zero counts.
+    """
+    az, el, _ = cartesian_to_spherical(v)
+
+    # 3D binning.
+    hist, _ = np.histogramdd(
+        sample=(az, el, energy), bins=[az_bin_edges, el_bin_edges, energy_bin_edges]
+    )
+
+    non_zero_counts = hist[hist > 0]
+
+    return hist, non_zero_counts
+
+
+def create_unique_identifiers(
```

Comment:
I don't understand what this function does. Is the point that you only want to store data in the file for spatial bins that have counts in them at one or more energy levels? So your counts array in the CDF is shaped like `n_b x n_e `where `n_b` is the number of spatial bins with at least one count in at least one energy bin and `n_e` is the number of energy bins?

Reply:
hist.size = 6220800 and that doesn't yet include the bin for species. So what I am doing is eliminating the bins that do not have counts.

This is where I get the indices for the nonzero bins:
az_bin_id, el_bin_id, energy_bin_id = np.argwhere(hist > 0).T
az_bin_id = array([150, 269, 269])
el_bin_id = array([99, 78, 78])
energy_bin_id=array([ 0,  1, 21])

We have 4 counts from our test data. We can use these indices to find the values in the histogram:
hist[150,99,0]
np.float64(1.0)
hist[269,78,1]
np.float64(1.0)
hist[269,78,21]
np.float64(2.0)

But as you can see the bins are sometimes the same so we just need to know the unique ones:
bin_id = (array([150, 269]), array([78, 99]), array([ 0,  1, 21]))

Then we have to map the original bin indices to the unique bin indices:
az_bin_id
array([150, 269, 269])
bin_id[0]
array([150, 269])
az_positions
array([0, 1, 1])

We know that the counts will have the dimension of these unique ids:
counts = np.zeros((len(bin_id[0]), len(bin_id[1]), len(bin_id[2])))
counts[az_positions, el_positions, energy_positions] = non_zero_counts

az_positions
array([0, 1, 1])
el_positions
array([1, 0, 0])
energy_positions
array([0, 1, 2])
non_zero_counts
array([1., 1., 2.])

counts[0, 1, 0] = 1
counts[1, 0, 1] = 1
counts[1, 0, 2] = 2

So now we have
bin_id
(array([150, 269]), array([78, 99]), array([ 0,  1, 21]))
counts
array([[[0., 0., 0.],
        [1., 0., 0.]],
       [[0., 1., 2.],
        [0., 0., 0.]]])

And we can use it:
For the target bin combination [150, 99, 0]:
150 in bin_id[0] corresponds to index 0 because 150 is the first element.
99 in bin_id[1] corresponds to index 1 because 99 is the second element.
0 in bin_id[2] corresponds to index 0 because 0 is the first element.

counts[0, 1, 0] = 1.0